### PR TITLE
[CSS] Invalid @property rule should be ignored at parse-time

### PR DESCRIPTION
**Pull request by mdubet**
#### 31ae9755adc7c64da938e59c0c1eacc52d5636c7
<pre>
[CSS] Invalid @property rule should be ignored at parse-time
<a href="https://bugs.webkit.org/show_bug.cgi?id=263007">https://bugs.webkit.org/show_bug.cgi?id=263007</a>
rdar://116803886

Reviewed by Antti Koivisto and Tim Nguyen.

This patch moves most of the validity checks from rule-registration-time (during rule set building)
to parse-time, thus preventing invalid @property rules to appear in the CSSOM.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/at-property-cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/at-property-cssom.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/at-property-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/at-property.html:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumePropertyRule):
* Source/WebCore/style/CustomPropertyRegistry.cpp:
(WebCore::Style::CustomPropertyRegistry::registerFromStylesheet):

Canonical link: <a href="https://commits.webkit.org/269466@main">https://commits.webkit.org/269466@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/206425445e75a5c18bc40e2759afc3182dbcf3f2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22649 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/329 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23736 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24558 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20968 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22908 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1399 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23179 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21915 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22889 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19653 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25411 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/196 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20524 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26766 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20767 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24603 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/200 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18058 "Passed tests") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20327 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5390 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/213 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | |

### CODE REVIEW

File: Source/WebCore/css/parser/CSSParserImpl.cpp [hunk 1]
```diff
@@ -1040,25 +1040,54 @@ RefPtr<StyleRuleProperty> CSSParserImpl::consumePropertyRule(CSSParserTokenRange
...
+    if (!syntax->isUniversal()) {
+        if (!descriptor.initialValue)
+            return nullptr;
+    }
+
+    auto initialValueIsValid = [&] {
+        auto tokenRange = descriptor.initialValue->tokenRange();
+        auto dependencies = CSSPropertyParser::collectParsedCustomPropertyValueDependencies(*syntax, tokenRange, strictCSSParserContext());
+        return dependencies.isComputationallyIndependent();
+        // FIXME: We should also check values like "var(--foo)"
```
Comment (nt1m) [inline Source/WebCore/css/parser/CSSParserImpl.cpp:1085]: Is there a bug filed? It's unclear to me what this means.

**Review by nt1m — APPROVED** [review]

File: LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/at-property-cssom.html [hunk 2]
```diff
@@ -106,22 +120,26 @@
...
 
+// Invalid @property rules.
+test_invalid('--no-descriptors');
+test_invalid('--no-syntax');
+test_invalid('--no-inherits');
+test_invalid('--no-initial-color-value');
+test_invalid('--syntax-only', '@property --syntax-only { syntax: "<color> | none"; }');
+test_invalid('--inherits-only', '@property --inherits-only { inherits: true; }');
+test_invalid('--initial-value-only', '@property --initial-value-only { initial-value: red; }');
+
```
Comment (anttijk) [inline LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/at-property-cssom.html:131]: Please always do WPT update in a separate patch first.

File: Source/WebCore/style/CustomPropertyRegistry.cpp [hunk 3]
```diff
@@ -83,16 +83,8 @@ bool CustomPropertyRegistry::registerFromAPI(CSSRegisteredCustomProperty&& prope
...
 void CustomPropertyRegistry::registerFromStylesheet(const StyleRuleProperty::Descriptor& descriptor)
 {
-    // "The inherits descriptor is required for the @property rule to be valid; if it’s missing, the @property rule is invalid."
-    // https://drafts.css-houdini.org/css-properties-values-api/#inherits-descriptor
-    if (!descriptor.inherits)
-        return;
-
-    // "If the provided string is not a valid syntax string, the descriptor is invalid and must be ignored."
-    // https://drafts.css-houdini.org/css-properties-values-api/#the-syntax-descriptor
     auto syntax = CSSCustomPropertySyntax::parse(descriptor.syntax);
```
Comment (anttijk) [inline Source/WebCore/style/CustomPropertyRegistry.cpp:86]: I think StyleRuleProperty::Descriptor can now include syntax as CSSCustomPropertySyntax instead of a String. This way you won't have to parse again here.

File: LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/at-property-cssom.html (hunk 2 above)
Comment (nt1m) [reply LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/at-property-cssom.html:131]: For the sake of integration, I think this is OK.

**Review by anttijk — APPROVED** [review]

File: Source/WebCore/style/CustomPropertyRegistry.cpp (hunk 3 above)
Comment (maattdd) [reply Source/WebCore/style/CustomPropertyRegistry.cpp:86]: Indeed it's a bit annoying to double parse, but we need to keep the user-written syntax for serialization. https://drafts.css-houdini.org/css-properties-values-api/#serialize-a-csspropertyrule

Comment (mdubet) [reply Source/WebCore/style/CustomPropertyRegistry.cpp:86]: Indeed it's a bit annoying to double parse, but we need to keep the user-written syntax for serialization. https://drafts.css-houdini.org/css-properties-values-api/#serialize-a-csspropertyrule

File: Source/WebCore/css/parser/CSSParserImpl.cpp (hunk 1 above)
Comment (mdubet) [reply Source/WebCore/css/parser/CSSParserImpl.cpp:1085]: It needs some more exploration (cf the FAIL test) to correctly reject those.
